### PR TITLE
New semantic analyzer: Support Final

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes https://github.com/python/mypy/issues/6290

I also fix couple minor bugs necessary to enable existing test file and fix final-related errors in mypy self-check with new analyzer.

### CODE REVIEW

File: mypy/newsemanal/semanal.py [hunk 1]
```diff
@@ -1656,11 +1657,6 @@ def visit_import_from(self, imp: ImportFrom) -> None:
                     self.add_symbol(imported_id, gvar, imp)
                     continue
             if node and node.kind != UNBOUND_IMPORTED and not node.module_hidden:
-                if not node:
-                    # Normalization failed because target is not defined. Avoid duplicate
-                    # error messages by marking the imported name as unknown.
-                    self.add_unknown_symbol(imported_id, imp, is_import=True)
-                    continue
```
Comment (ilevkivskyi) [inline mypy/newsemanal/semanal.py]: This is dead code.

File: test-data/unit/pythoneval.test [hunk 2]
```diff
@@ -1365,7 +1365,8 @@ from typing import Dict, List, Tuple
 x: Dict[str, List[int]]
 reveal_type(x['test'][0])
 [out]
-typing.pyi: error: Class typing.Sequence has abstract attributes "__len__"
-typing.pyi: note: If it is meant to be abstract, add 'abc.ABCMeta' as an explicit metaclass
+typing.pyi:251: error: Class typing.Sequence has abstract attributes "__len__"
+typing.pyi:251: note: If it is meant to be abstract, add 'abc.ABCMeta' as an explicit metaclass
+ast.pyi:31: error: Name 'PyCF_ONLY_AST' already defined (possibly by an import)
```
Comment (ilevkivskyi) [inline test-data/unit/pythoneval.test]: This is an actual bug in typeshed, this is name is indeed already defined by a star import.

File: mypy/newsemanal/semanal.py [hunk 3]
```diff
@@ -3937,7 +3945,7 @@ def is_global_or_nonlocal(self, name: str) -> bool:
                      or name in self.nonlocal_decls[-1]))
 
     def add_symbol_table_node(self, name: str, symbol: SymbolTableNode,
-                              context: Optional[Context] = None) -> None:
+                              context: Optional[Context] = None) -> bool:
         """Add symbol table node to the currently active symbol table."""
```
Comment (JukkaL) [inline mypy/newsemanal/semanal.py]: Document the return value here as well.

File: mypy/newsemanal/semanal.py [hunk 4]
```diff
@@ -4035,18 +4045,23 @@ def name_not_defined(self, name: str, ctx: Context, namespace: Optional[str] = N
...
+        else:
+            node = None
 
         if isinstance(original_ctx, SymbolTableNode) and isinstance(original_ctx.node, MypyFile):
             # Since this is an import, original_ctx.node points to the module definition.
             # Therefore its line number is always 1, which is not useful for this
             # error message.
             extra_msg = ' (by an import)'
-        elif node and node.line != -1:
+        elif node and node.line != -1 and node.fullname().startswith(self.cur_mod_id):
```
Comment (JukkaL) [inline mypy/newsemanal/semanal.py]: A `startswith` check is not reliable. For example, `foo` is a prefix of `foobar`, even though they are unrelated names.

**Pull request merged**